rfc/plugins.md

```diff
@@ -0,0 +1,343 @@
+
+### Deployment of Remote Plugins
+Remote plugins can be deployed using whatever means the plugin developer desires.
+In the case of a remote plugin being developed by a third-party outside of the organization, the operator may need to deploy and maintain the plugin themselves.
+Halyard can be used for registering remote plugins with Front50.
+In the absence of Halyard, registering remote plugins with Front50 would be done through the API, just as Halyard would do.
+
+### Deployment of In-Process Plugins
+Downloading plugin resources is necessary for in-process plugins.
+Services need to have plugin resources available locally in order to load them into their process.
+Plugin resources are hosted on a remote server and must be downloaded in order to use them.
+The existing method of downloading plugin resources involves the service itself downloading plugins via a given URI.
+
+#### Kubernetes Environment
+Halyard can be used for deploying plugins to Kubernetes environments.
+To deploy plugins, Halyard will follow the same pattern as Kubernetes regarding downloading resources prior to runtime.
+Downloading of resources will be done through init containers and volume mounts that are then shared with their respective service.
+By using init containers to download resources, we can standardize on one method of deploying plugins to spinnaker services.
+Plugin resources will be placed in the following locations:
+
+- Java services: `/opt/spinnaker/plugins/resources`
+- Deck: `/opt/deck/html/plugins`
+
+Plugins are placed in a different location for Deck so that it can serve plugin resources.
+
+#### VM Based Environment
+Plugins and their configurations can be downloaded or baked into the image.
+
+### Plugin Configuration
+
+#### Remote Plugins
+Halyard or the operator, depending on preference, will be tasked with providing plugin configurations and plugin overrides to the remote plugins.
+
+#### In-Process Plugins
+PF4J plugins are configured with properties files and deployments are configured via `manifest.yml` file.
+Properties files configure the application and `manifest.yml` lets halyard know where plugin resources are located.
+`manifest.yml` can also override properties from properties file.
+PF4J will handle loading plugins into a Spinnaker service, while `manifest.yml` will inform which resources should be deployed to each service.
+
+### Dependencies
+A series of new kork modules will be provided for libraries and services alike to consume to expose new extension points. 
+Under the covers, this introduces PF4J as a new global service dependency.
+
+In-process plugins, by default, will have their own ClassLoader that is isolated from other Plugin ClassLoaders. 
+These Plugin ClassLoaders will be children of the root ClassLoader, meaning they cannot have dependencies that conflict with the root ClassLoader. 
```

Comment:
> they cannot have dependencies that conflict with the root ClassLoader

Given the sheer quantity of jars used by Spinnaker, this is going to be nearly impossible. We can't even always get this right inside a given Spinnaker service; imagine how hard it's going to be for a plugin. You think you can just tell everyone to use the dependencies from `kork` but either of these very common scenarios will screw that up:
1) they're going to add dependencies of their own (since they'll need additional jars to add functionality) which will have transitive dependencies and, say, upgrade the version of Guava without realizing it
2) they build the plugin against Spinnaker 1.16.3 but try to run it in 1.17.0

I've worked on several products that used plugins (most recently IntelliJ) and I can tell you this is guaranteed to cause issues that are super hard to reason about (but are at least generally trivial to reproduce).

I think what you want to do is this:
1) extension points are built against a super minimal classpath... ideally just the JDK and some simple data classes. Maybe a library like `protobuf` that's specifically designed to have no other downstream dependencies.
2) plugins are built against this extension point but don't include those classes in their distribution jar
3) the root classloader for plugins has _only_ the jars from the extension point and its (minimal) dependencies

If that makes sense to you, I can try to add some documentation about this as a follow-up PR?

Reply:
Yes that makes sense. In fact that's what we were landing on (see the latest dependency cleanup PR I have in kork, and will follow-up with cleaning up orca-api as well).

It's also why we want to standardize around protobuf as the contract, to help avoid this class of problems - definitely agree it'll be a headache.

Would love if you could follow-up PR on this, thank you!

Reply:
This feels like the thing to get (more) correct initially (vs. the earlier discussion around plugin registries).

@plumpy's suggestion makes sense to me.